sympy/core/containers.py

```diff
@@ -264,3 +266,53 @@ def __lt__(self, other):
     def _sorted_args(self):
         from sympy.utilities import default_sort_key
         return tuple(sorted(self.args, key=default_sort_key))
+
+
+class OrderedSet(collections.MutableSet):
+    def __init__(self, iterable=None):
+        if iterable:
+            self.map = collections.OrderedDict((item, None) for item in iterable)
+        else:
+            self.map = collections.OrderedDict()
+
+    def __len__(self):
+        return len(self.map)
+
+    def __contains__(self, key):
+        return key in self.map
+
+    def add(self, key):
+        self.map[key] = None
+
+    def discard(self, key):
+        self.map.pop(key)
+
+    def pop(self, last=True):
+        return self.map.popitem(last=last)[0]
+
+    def __iter__(self):
+        for key in self.map.keys():
+            yield key
+
+    def __repr__(self):
+        if not self.map:
+            return '%s()' % (self.__class__.__name__,)
+        return '%s(%r)' % (self.__class__.__name__, list(self.map.keys()))
+
+    def intersection(self, other):
+        result = []
+        for val in self:
```

Comment:
Suggestion: The outer loop look should through the smaller of the two sets to reduce the number of lookups that are needed.

Reply:
Never mind, that does not preserve the order.